lib/ref_arch_setup/install.rb

```diff
@@ -70,6 +78,223 @@ def handle_pe_conf(pe_conf_path)
+
+    # Runs the download_pe_tarball Bolt task
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The pe tarball URL
+    # @param [string] nodes Nodes where the task should be run
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def download_pe_tarball(url, nodes)
+      puts "Attempting to download #{url} to #{nodes}"
+      puts
+
+      params = {}
+      params["url"] = url
+      params["destination"] = TMP_WORK_DIR
+
+      success = BoltHelper.run_task_with_bolt("ref_arch_setup::download_pe_tarball", params, nodes)
+      return success
+    end
+
+    # Downloads the PE tarball locally and moves it to the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The pe tarball URL
+    # @param [string] filename The pe tarball filename
+    # @param [string] target_master Host to install on
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def download_and_move_pe_tarball(url, filename, target_master)
+      puts "Attempting to download #{url} to localhost and move to #{target_master}"
+      puts
+
+      success = download_pe_tarball(url, "localhost")
+      success = upload_pe_tarball("#{TMP_WORK_DIR}/#{filename}", target_master) if success
+      return success
+    end
+
+    # Downloads the specified PE tarball URL by either downloading directly to the
+    # target master or downloading locally and copying to the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The pe tarball URL
+    # @param [string] target_master Host to install on
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def handle_tarball_url(url, target_master)
+      uri = URI.parse(url)
+      filename = File.basename(uri.path)
+
+      # TODO: improve check for localhost
+      if target_master.equal?("localhost")
+
+        # destination = default
+        success = download_pe_tarball(url, target_master)
+        raise "Failed downloading #{url} to #{target_master}" unless success
+
+      else
+
+        puts "Specified target master #{target_master} is not localhost"
+        puts
+
+        success = download_pe_tarball(url, target_master)
+
+        # if at first you don't succeed...
+        puts "Failed download attempt to #{target_master}." unless success
+        success = download_and_move_pe_tarball(url, filename, target_master) unless success
+
+        raise "Failed downloading #{url} to localhost and moving to #{target_master}" unless success
+
+      end
+
+      return filename
+    end
+
+    # Copies the PE tarball from the specified location to the temp working directory
+    # on the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] pe_tarball_path The pe tarball path
+    # @param [string] target_master Host to install on
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def copy_pe_tarball(pe_tarball_path, target_master)
```

Comment:
This essentially has the requirement that the path and target master be for the same host.  But it doesn't enforce that.  That might be okay though.  Just feels odd.